### DISCUSSION TITLE
Reorganizing folders to a python project

### DISCUSSION THREAD

**Discussion by brando90** [princeton-vl/CoqGym#38] [Q&A]
Hi Kaiyu!

What do you think of organizing the code base as a standard python package so people interested in your work can more easily install the great code for your project?

Usually what I do is put everything in a `repo-name-proj` folder. Then inside it a `setup.py` as follows:

```
from setuptools import setup
from setuptools import find_packages

setup(
    name='packaging_project',  # project name
    version='0.1.0',
    description="Brando's sample packaging tutorial",
    #url
    author='Brando Miranda',
    author_email='miranda9@illinois.edu',
    license='MIT',
    packages=find_packages(),  # default find_packages(where='.', exculde=())
    install_requires=['torch', 'numpy', 'scikit-learn', 'scipy', 'matplotlib', 'pyyml', 'torchviz', 'tensorboard',
    'graphviz', 'torchvision', 'matplotlib']
)
```

then people can use `pip install` and editable mode installs (the nice one advantage!) like

```
pip install -e .
conda develop .
```

in the location where `setup.py` is.

Let me know what you think! :)

----
References:

- https://packaging.python.org/tutorials/packaging-projects/
- http://code.nabla.net/doc/setuptools/api/setuptools/setuptools.find_packages.html

**Comment by yangky11** [comment]
Hi Brando,

Does this handle OPAM packages, SerAPI, CoqHammer, etc. (The steps 1–5 in the README). Or is it only for Python?

**Reply by brando90** [reply]
Hi Kaiyu,

This is only for packaging python. Let me know what you think and how I can help.

**Reply by brando90** [reply]
> Hi Brando,
> 
> Does this handle OPAM packages, SerAPI, CoqHammer, etc. (The steps 1–5 in the README). Or is it only for Python?

for example you won't have to do things like this anymore:
```
sys.path.append(os.path.sep.join(__file__.split(os.path.sep)[:-2]))
```

since you'd `conda develop  .` your python projects in your environment :)

**Reply by yangky11** [reply]
I see. I haven't used Python packaging before. You are welcome to submit a preliminary pull request and we can take a look together and start from there.

**Reply by brando90** [reply]
@yangky11
 
I will do that!

----

also, slightly unrelated but I thought you'd love this (I did when I discovered it!).

Python 3 has a fantastic way to format strings using f-string (https://realpython.com/python-f-strings/) e.g.
```
print(f' epoch number: {n_epoch}')
```

also the pathlib library for managing folder manipulation etc is beautiful too. My favorite part is to manage things system independent e.g.
```
from pathlinb import path
sexp_cache_dir = Path('~/data/coqgym/sexp_cache').expanduser()
sexp_cache_dir.mkdir(parents=True, exist_ok=True)
```
hope you like these! :)

**Reply by yangky11** [reply]
It's great to know. Thanks! I'm more used to using `%d`, `%s`, `%f` style string formatting, maybe because the first programming language i learned was C. But the new way in Python is cool!

**Reply by brando90** [reply]
> It's great to know. Thanks! I'm more used to using `%d`, `%s`, `%f` style string formatting, maybe because the first programming language i learned was C. But the new way in Python is cool!

also I forgot the pathlib library allows adding folder paths in an intelligent way with operators:

```
path = Path(...) / p1 / p2 / filename
``` 

:D
